### ISSUE TITLE
Provide an DB upgrade script from 9.2.5 to 9.3.3

### ISSUE THREAD

**Issue by bourtemb** [tango-controls/TangoSourceDistribution#41]
Hello,

this is juste that dbconfig-common does an incremental upgrade from on version to the other.
So I just need a script whcih does the migration from 9.2.5 to 9.3.3

DELETE FROM property_class WHERE class='DServer' AND count >= 11;
INSERT INTO property_class VALUES('DServer','AllowedAccessCmd',11,'EventConfirmSubscription',NOW(),NOW(),NULL);

Is it the only change  ?

_Originally posted by @picca in https://github.com/tango-controls/TangoSourceDistribution/pull/18#issuecomment-498024593_

**Comment by mliszcz** [comment]
Hi @picca,
I need your help to clarify what is needed here.

My questions are:
1. do you need something like existing `update_db8.sql.in`, called e.g. `update_db9.2.5.sql.in`?
2. if yes, then the script shall probably go to TangoDatabase project - @bourtemb could you confirm?

I tried to look into debian package sources to find existing scripts for incremental upgrades, but I couldn't find anything (e.g. script for upgrading 9.2.1 to 9.2.5).

There is no difference (with @TANGO_DB_NAME@ = tango) between:
https://salsa.debian.org/science-team/tango/blob/master/debian/dbconfig-common/data/tango-db/upgrade/mysql/9.2.5
and:
https://github.com/tango-controls/TangoDatabase/blob/master/stored_proc.sql.in

No new procedures were added comparing to:
https://salsa.debian.org/science-team/tango/blob/master/debian/dbconfig-common/data/tango-db/upgrade-dbadmin/mysql/9.2.5

You have `EventConfirmSubscription` configured in:
https://salsa.debian.org/science-team/tango/blob/master/cppserver/database/create_db.sql.in#L136

**Comment by bourtemb** [comment]
>     1. do you need something like existing `update_db8.sql.in`, called e.g. `update_db9.2.5.sql.in`?
> 
>     2. if yes, then the script shall probably go to TangoDatabase project - @bourtemb could you confirm?

That would be the best location to do that indeed.

> I tried to look into debian package sources to find existing scripts for incremental upgrades, but I couldn't find anything (e.g. script for upgrading 9.2.1 to 9.2.5).
> 
> There is no difference (with @TANGO_DB_NAME@ = tango) between:
> https://salsa.debian.org/science-team/tango/blob/master/debian/dbconfig-common/data/tango-db/upgrade/mysql/9.2.5
> and:
> https://github.com/tango-controls/TangoDatabase/blob/master/stored_proc.sql.in
> 
> No new procedures were added comparing to:
> https://salsa.debian.org/science-team/tango/blob/master/debian/dbconfig-common/data/tango-db/upgrade-dbadmin/mysql/9.2.5

Indeed. No new stored procedure since 9.2.5.

> 
> You have `EventConfirmSubscription` configured in:
> https://salsa.debian.org/science-team/tango/blob/master/cppserver/database/create_db.sql.in#L136

Indeed, but this is coming from latest 9.3.3rc1 source distrib, not from 9.2.5.

Here are the differences between 9.2.5 and 9.3.3:
https://salsa.debian.org/science-team/tango/commit/9f671cf375d9a06240f7f6b73d5f883dc5dbb0c3#2f1bd2e8ec89fb0b71be7fa082f997a5c21266f2

So, EventConfirmSubscription command was added to the list of always allowed commands in the TAC and the declaration of the rest server was added.

**Comment by mliszcz** [comment]
I have opened a PR in TangoDatabase: https://github.com/tango-controls/TangoDatabase/pull/18

Once that one is merged, we need to create a new tag in TangoDatabase and update source distribution.

@picca I am unable to add you as a reviewer. Please tell whether the solution from above mentioned PR is ok for you.

**Comment by picca** [comment]
I think that this is ok.

I just remember, (but it is a long time ago, that we have to deal also with the sdtored procedures :).
I do not remmermber exactly how I did, but it seems to me that we decided to remove them completly and regenerate the procedures at each upgrade.

**Comment by picca** [comment]
Here the target I was using in order to produce the upgrade scripts. before each release of the package.
where VERSION is the version of tango, with the latest 9.2.5
Neverthless I need to change this target for each release. sometimes there is uopgrade scripts and sometimes there is nothing bu the procedures.

So at the end I merged the upgrade scripts and the procedures.


It would be nice if we could just take the upstream migration script and put it in the packaging as it.
It would be great also for other distro.


cheers

sql:
	./configure $(CONFIGURE_OPTS)
	# use this target to generate the dbconfig-common upgrade script once the debian/changelog is ok
	echo "USE tango;\n" > debian/dbconfig-common/data/tango-db/upgrade/mysql/$(VERSION)
	cat cppserver/database/stored_proc.sql >> debian/dbconfig-common/data/tango-db/upgrade/mysql/$(VERSION)
	$(MAKE) distclean

**Comment by mliszcz** [comment]
Hi @picca,
Thanks for the comment.
I've sourced `stored_proc.sql` from `update_db9.2.5.sql`.
Hope it is ok now and you can use this script directly in debian package.
// EDIT: the PR in TangoDatabase is updated

**Comment by picca** [comment]
> Hi @picca<https://github.com/picca>,
> Thanks for the comment.
> I've sourced stored_proc.sql from update_db9.2.5.sql.
> Hope it is ok now and you can use this script directly in debian package.

I need only one file this is why I merge the files all together. :). It is easier.
but maybe we can decide to put the stored procedures somewhere on /usr/share/tango/db/stored_procedures/store_x.y.z.sql and source it during the upgrade.
but this way the upgrade script could be generated during the configuretion in order to use the right path  for th e stored_procedure.
the stored procedure should be versionned also and keep from one package to the other.

We need to all migration from  all the tango version distributed by Debian to the latest on.

cheers

**Comment by mliszcz** [comment]
So, we already have a script for Tango 9.2.5 (update_db.9.2.5.sql for stretch) and for Tango 8 (update_db8.sql for jessie).
Is this enough or some more work is needed?

**Comment by picca** [comment]
I do not thingk so, because I put all these scripts  in the packaging here[1]
As you see now I just need a 9.3.3 file :)

[1] https://sources.debian.org/src/tango/9.2.5a+dfsg1-2/debian/dbconfig-common/data/tango-db/upgrade/mysql/

**Comment by mliszcz** [comment]
For 9.3.3 you should use `update_db.9.2.5.sql` which is being added to TangoDatabase. Is anything missing?

**Comment by picca** [comment]
> For 9.3.3 you should use update_db.9.2.5.sql which is being added to TangoDatabase. Is anything missing?

maybe the update script should be named update_db_from.9.2.5.to.9.3.3.sql to be completly explicite :))
this file source the procedure, but I do not know what is the path of the 9.3.3 procedure sql file, when my package is installed.

the best for me is to have only one file  which does eveythings without sourcing external files whcih are not necessarely install where you think they are :).
this is why I do a cat in my debian rules.

now we can decide what is the best way to produce this uniq file during the build :)

**Comment by mliszcz** [comment]
> maybe the update script should be named update_db_from.9.2.5.to.9.3.3.sql to be completly explicite :))

I tried to use existing naming schema which is `update_dbXXX.sql` to upgrade from XXX to latest.
1. I think we would need to rename other files as well for consistency.
2. We'd need to decide what to do on next release - provide a script just for 9.3.3 to 9.x.x upgrade?

> but I do not know what is the path of the 9.3.3 procedure sql file

On my system (installed from source distribution) it is: /usr/local/share/tango/db/stored_proc.sql

> the best for me is to have only one file which does eveythings without sourcing external files

stored_proc.sql is 1100 lines long. If included in every migration script, we will have a lot of duplicated code. This is bad for maintainability.

It would be good if other kernel developers could share their opinions on how we'd like to organize migration scripts. In my opinion, on each stable release we could provide migration script only from previous stable release.

@bourtemb can you comment on above points?

Also, what needs clarification is why are we re-creating all the procedures even if there were no change since last release? @picca could you try to find any reference (forum post, email, etc.) for this?:
> I just remember, (but it is a long time ago, that we have to deal also with the sdtored procedures :). I do not remmermber exactly how I did, but it seems to me that we decided to remove them completly and regenerate the procedures at each upgrade.

**Comment by Ingvord** [comment]
> In my opinion, on each stable release we could provide migration script only from previous stable release.

+1. In fact this is how it usually done. If someone needs to upgrade from older versions, they apply scripts one by one

**Comment by bourtemb** [comment]
> In my opinion, on each stable release we could provide migration script only from previous stable release.
> 
> +1. In fact this is how it usually done. If someone needs to upgrade from older versions, they apply scripts one by one

I think this is the way we should go for the new versions. For each new stable release, provide a new migration script from previous stable release to the new one. But we should keep the older scripts too so a user coming from Tango 8 should be able to update to the previous stable version first, and then apply the latest migration script from the previous version to the latest stable version.

**Comment by mliszcz** [comment]
Ok, I'll rename the script to `update_db_from_9.2.5_to_9.3.3.sql`.
@Ingvord, @bourtemb do you know if we need to recreate the procedures during migration? Maybe when mysql is upgraded together with tango ...

**Comment by bourtemb** [comment]
We don't need to recreate the procedures during the migration.
For instance, when migrating from 9.2.5 to 9.3.3, the stored procedures didn't change at all.
The migration is for the use case where the user has already a running Tango DB. So the stored procedures are already there.
When migrating from older tango versions to Tango 9.2.5 or Tango 9.3.3, the stored procedures need to be updated. They don't need to be recreated, the migration script from Tango 9.x to 9.2.5 will only recreate the ones which changed.

**Comment by mliszcz** [comment]
@bourtemb please see [`stored_proc.sql.in`](https://github.com/mliszcz/TangoDatabase/blob/master/stored_proc.sql.in) which is sourced from all upgrade scripts. First we drop all the procedures and then create them again. @picca pointed out that there were some problems in the past and re-creating the procedures helped to resolve these issues.

**Comment by bourtemb** [comment]
Well, it depends on what kind of issues they were encountering in the past...
The thing is that in the past, the stored procedures had changed from one Tango release to another.
From 9.2.5 to 9.3.3, they didn't change.
The stored procedures are saved in a database named mysql, which is a kind of mysql system table.
There might be some issues with this table when upgrading MySQL version at the same time as tango, which is usually the case when upgrading the Linux distribution?
And also, there might be some issues when moving from MySQL to mariaDB.
This might be the reason why the stored procedures are systematically recreated when migrating?

**Comment by mliszcz** [comment]
Thanks @bourtemb,
@picca I've checked mysql's documentation for in-place binary upgrades.
https://dev.mysql.com/doc/refman/5.7/en/upgrade-binary-package.html
https://dev.mysql.com/doc/refman/8.0/en/upgrade-binary-package.html
https://dev.mysql.com/doc/refman/8.0/en/upgrading-what-is-upgraded.html

Per my understanding, whole mysql schema with all procedures (data dictionary) shall be upgraded automatically.

> When you start the MySQL 8.0 server, it automatically detects whether data dictionary tables are present. If not, the server creates them in the data directory, populates them with metadata, and then proceeds with its normal startup sequence. **During this process, the server upgrades metadata for all database objects, including databases, tablespaces, system and user tables, views, and stored programs (stored procedures and functions, triggers, and Event Scheduler events).** The server also removes files that previously were used for metadata storage. For example, after upgrading from MySQL 5.7 to MySQL 8.0, you will notice that tables no longer have .frm files.

I suggest to NOT source stored_proc.sql from 9.2.5 to 9.3.3 migration script, i.e. my last chage has to be reverted. Would it be ok for you?

**Comment by picca** [comment]
ok for me.


the last time I had a problem with the ownership of the procedure this is why I had to regenerate them.
only the tango user can access these procedures.

Cheers

**Comment by bourtemb** [comment]
> I suggest to NOT source stored_proc.sql from 9.2.5 to 9.3.3 migration script, i.e. my last chage has to be reverted. Would it be ok for you?

OK for me too.

**Comment by mliszcz** [comment]
Changes applied to tango-controls/TangoDatabase#18:
* TangoRestServer added to existing upgrade scripts (entry in `devices` table was not created e.g. when upgrading from Tango 8)
* all `update_dbX.sql.in` renamed to `update_db_from_X_to_9.3.3.sql.in`
* `update_db_from_9.2.5_to_9.3.3.sql.in` added as discussed in this issue

**Comment by bourtemb** [comment]
In the next versions, we won't update anymore the update_db_from_X_to_9.3.3.sql.in scripts and we will only provide 1 additional new script to update from the previous version to the next version.
A user updating from version X to version Z will have to execute the update_db_from_X_to_9.3.3.sql.in and then all the scripts for the intermediate versions (update_db_from_9.3.3_to_Y.sql.in and update_db_from_Y_to_Z.sql.in).

**Comment by picca** [comment]
+1 :))

**Comment by mliszcz** [comment]
I have to reopen this one. I used 'resolve' keyword in TangoDatabase PR and this issue was closed accidentally.

Another PR is needed for TangoSourceDistribution to copy the new scripts during installation. I'll work on this.

**Comment by picca** [comment]
Hello,

where do you want to install them ?
At the end the migration scripts need to be renamed and install in a dedicated Debian specific location. I do not know how other distribution are working on this :).

Cheers

**Comment by mliszcz** [comment]
Hi @picca 
The new script is installed next to the old scripts, in `share/tango/db`. This is the default location when installing from the source tarball. Nothing  debian-specific.

```
$ ll /prefix/share/tango/db/
total 100
drwxr-xr-x 2 root root  4096 sie  6 09:19 ./
drwxr-xr-x 4 root root  4096 sie  6 09:19 ../
-rw-r--r-- 1 root root   898 sie  6 09:19 create_db.sh
-rw-r--r-- 1 root root  9702 sie  6 09:19 create_db.sql
-rw-r--r-- 1 root root 10063 sie  6 09:19 create_db_tables.sql
-rw-r--r-- 1 root root   610 sie  6 09:19 rem_history.sql
-rw-r--r-- 1 root root 32760 sie  6 09:19 stored_proc.sql
-rw-r--r-- 1 root root  5821 sie  6 09:19 update_db_from_7_to_9.3.3.sql
-rw-r--r-- 1 root root  5376 sie  6 09:19 update_db_from_8_to_9.3.3.sql
-rw-r--r-- 1 root root   721 sie  6 09:19 update_db_from_9.2.5_to_9.3.3.sql
-rw-r--r-- 1 root root   482 sie  6 09:19 update_db.sh
-rw-r--r-- 1 root root    74 sie  6 09:19 update_db.sql
```

**Comment by picca** [comment]
thanks, in my case, I will generate a migration script called `9.3.3` and put everything in it in order to do the migration from 9.2.5 to 9.3.3. I will take as base  `update_db_from_9.2.5_to_9.3.3.sql`.
I need to replace all the source directive by an embeded copy in orde rto have an autoportant script.

I did this choice, because the stored_proc.sql file is not versionned. So when upgrading from tnago7 to tango 9.3.3, I can not source the stored_procedure_7.sql then the stored_procedure_8 etc...

maybe the best would be to version also these procedures.

Cheers

**Comment by mliszcz** [comment]
Hi @picca,
I think we will not touch `stored_proc.sql` anymore. If a procedure is updated in the next release (9.x) we will alter it directly from `update_db_from_9.3.3_to_9.x.sql`.

E.g. the upgrade path from Tango 7 will be:
1. `update_db_from_7_to_9.3.3.sql` (which sources `stored_proc.sql`)
2. `update_db_from_9.3.3_to_9.x.sql` (which modifies procedures created by `stored_proc.sql`)

Update:
Also note that `update_db_from_9.2.5_to_9.3.3.sql` does not source anything. You can just rename it to '9.3.3' without further modifications.

**Comment by bourtemb** [comment]
@picca, the stored_proc.sql.in file drops all the stored procedures and recreate them, ensuring you always get the latest version of the stored procedures when executing it.
Then, the scripts update_db_from_9.3.3_to_9.x.sql will take care of invoking stored_proc.sql or not if needed.
In your case, you don't need to worry about it. You will just need to invoke the update_db_from_a.b.c_to_x.y.sql script corresponding to the right versions.

> Also note that `update_db_from_9.2.5_to_9.3.3.sql` does not source anything. You can just rename it to '9.3.3' without further modifications.

I think it would not be correct to rename it. It does not source anything, precisely because the user is updating from 9.2.5. If you update from an older Tango version, you would need to update the stored procedures. I think it is important to keep 9.2.5 in the name to avoid confusion.

**Comment by bourtemb** [comment]
> thanks, in my case, I will generate a migration script called `9.3.3` and put everything in it in order to do the migration from 9.2.5 to 9.3.3. I will take as base `update_db_from_9.2.5_to_9.3.3.sql`.
> I need to replace all the source directive by an embeded copy in orde rto have an autoportant script.
> 
> I did this choice, because the stored_proc.sql file is not versionned. So when upgrading from tnago7 to tango 9.3.3, I can not source the stored_procedure_7.sql then the stored_procedure_8 etc...

stored_proc.sql is versioned (in git). I don't understand what is the problem with this file.
In a normal installation. stored_proc.sql provided with the package will always correspond to the stored procedures of the Tango version of the package.

**Comment by picca** [comment]
ok, thanks all of you, for the clarification :))

Fred
